File: project_src/dfb_reinforcement_learning/policy_contract.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any


POLICY_CONTRACT_PATH = (
    Path(__file__).resolve().parents[2]
    / "config"
    / "dfb_reinforcement_learning"
    / "part3_policy_contract_v1.json"
)


def _load_policy_contract() -> tuple[bytes, dict[str, Any]]:
    try:
        raw = POLICY_CONTRACT_PATH.read_bytes()
    except OSError as exc:
        raise RuntimeError(f"unable to read policy contract: {POLICY_CONTRACT_PATH}") from exc
    try:
        payload = json.loads(raw)
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise RuntimeError(f"invalid policy contract JSON: {POLICY_CONTRACT_PATH}") from exc
    if not isinstance(payload, dict):
        raise RuntimeError("policy contract root must be an object")
    return raw, payload


POLICY_CONTRACT_BYTES, POLICY_CONTRACT = _load_policy_contract()
POLICY_CONTRACT_SHA256 = hashlib.sha256(POLICY_CONTRACT_BYTES).hexdigest()

POLICY_CONTRACT_ID = str(POLICY_CONTRACT["policy_contract_id"])
OBSERVATION_SCHEMA_ID = str(POLICY_CONTRACT["observation"]["schema_id"])
ACTION_SCHEMA_ID = str(POLICY_CONTRACT["action"]["schema_id"])
NORMALIZER_SCHEMA_ID = str(POLICY_CONTRACT["normalizer_schema_id"])
DATASET_SCHEMA_ID = str(POLICY_CONTRACT["dataset_schema_id"])
CHECKPOINT_SCHEMA_ID = str(POLICY_CONTRACT["checkpoint_schema_id"])
MODEL_FAMILY_ID = str(POLICY_CONTRACT["model_family_id"])
OBS_DIM = int(POLICY_CONTRACT["observation"]["dim"])
ACTION_CONT_DIM = len(POLICY_CONTRACT["action"]["continuous"])
ACTION_BIN_DIM = len(POLICY_CONTRACT["action"]["binary"])

File: project_src/dfb_game/src/policy_contract.rs
use sha2::{Digest, Sha256};

pub const POLICY_CONTRACT_ID: &str = "dfb_part3_policy_contract_v1";
pub const OBSERVATION_SCHEMA_ID: &str = "dfb_part3_policy_observation_v1";
pub const ACTION_SCHEMA_ID: &str = "dfb_part3_policy_action_v1";
pub const NORMALIZER_SCHEMA_ID: &str = "dfb_part3_policy_normalizer_v1";
pub const DATASET_SCHEMA_ID: &str = "dfb_part3_policy_dataset_v1";
pub const CHECKPOINT_SCHEMA_ID: &str = "dfb_part3_policy_checkpoint_v1";
pub const MODEL_FAMILY_ID: &str = "dfb_part3_stateless_hybrid_actor_critic_v1";
pub const OBS_DIM: usize = 69;
pub const ACTION_CONT_DIM: usize = 4;
pub const ACTION_BIN_DIM: usize = 3;
pub const BINARY_OBS_INDICES: [usize; 10] = [22, 24, 26, 27, 29, 45, 59, 61, 62, 64];

pub const CONTRACT_BYTES: &[u8] =
    include_bytes!("../../../config/dfb_reinforcement_learning/part3_policy_contract_v1.json");

pub fn policy_contract_sha256() -> String {
    hex::encode(Sha256::digest(CONTRACT_BYTES))
}
